editors/code/src/installation/language_server.ts

```diff
@@ -0,0 +1,119 @@
+                });
+
+                lastPrecentage = newPercentage;
+            });
+        }
+    );
+
+    await fs.chmod(installationPath, 111); // Set xxx permissions
+}
+export async function ensureLanguageServerBinary(
+    langServerSource: null | BinarySource
+): Promise<null | string> {
+
+    if (!langServerSource) {
+        vscode.window.showErrorMessage(
+            "Unfortunately we don't ship binaries for your platform yet. " +
+            "You need to manually clone rust-analyzer repository and " +
+            "run `cargo xtask install --server` to build the language server from sources. " +
+            "If you feel that your platform should be supported, please create an issue " +
+            "about that [here](https://github.com/rust-analyzer/rust-analyzer/issues) and we " +
+            "will consider it."
+        );
+        return null;
+    }
+
+    switch (langServerSource.type) {
+        case BinarySourceType.ExplicitPath: {
+            if (isBinaryAvailable(langServerSource.path)) {
+                return langServerSource.path;
+            }
+            vscode.window.showErrorMessage(
+                `Unable to execute ${'`'}${langServerSource.path} --version${'`'}. ` +
+                "To use the bundled language server, set `rust-analyzer.raLspServerPath` " +
+                "value to `null` or remove it from the settings to use it by default."
+            );
+            return null;
+        }
+        case BinarySourceType.GithubBinary: {
+            const bundledBinaryPath = path.join(langServerSource.dir, langServerSource.file);
+
+            if (!isBinaryAvailable(bundledBinaryPath)) {
+                const userResponse = await vscode.window.showInformationMessage(
+                    `Language server binary for rust-analyzer was not found. ` +
+                    `Do you want to download it now?`,
+                    "Download now", "Cancel"
+                );
+                if (userResponse !== "Download now") return null;
+
+                try {
+                    await downloadLatestLanguageServer(langServerSource);
+                } catch (err) {
+                    await vscode.window.showErrorMessage(
+                        `Failed to download language server from ${langServerSource.repo.name} ` +
+                        `GitHub repository: ${err.message}`
+                    );
+                    return null;
+                }
+
+
+                assert(
+                    isBinaryAvailable(bundledBinaryPath),
+                    "Downloaded language server binary is not functional"
+                );
+
+                vscode.window.showInformationMessage(
+                    "Rust analyzer language server was successfully installed"
+                );
+            }
+            return bundledBinaryPath;
+        }
+    }
+
+    function isBinaryAvailable(binaryPath: string) {
+        return spawnSync(binaryPath, ["--version"]).status === 0;
```

Comment:
Maybe use the async version?

Reply:
I think I didn't use async version here because of extra parens that would be needed, but I don't have strong opinion on that...

Reply:
Yeah, I'm more used to stuff that runs on the UI thread (or appears to), where if you block the event loop the user might notice. Code extensions are run out of process, and you can't do much while the server is starting anyway, so it's not that bad here.

Reply:
Anyway, you are right, `async` function should be preferred to synchronous, this is the one of the main principles of JavaScript, so let's try not to break them

Reply:
@lnicola , well you know, actually, I did find the reason I left it synchronous.  It's because this `spawn()` function doesn't have `Promise` version. It uses some [legacy even-driven interface](https://nodejs.org/api/child_process.html#child_process_class_childprocess) that I don't want to dive into. Though I found [some `Promise` adapter on `npm`](https://github.com/patrick-steele-idem/child-process-promise/blob/217609b5e12f12616c1c3987fe7451b8fa293112/lib/index.js#L80-L152) for this specific function I don't think it is worth it anyway, so let's just leave it that simple for this case :man_facepalming:

Reply:
No worries, I might give it a try at some point. I have a feeling it's easy to adapt that pattern into a promise, so I might as well learn how.

Reply:
Extensions are run in a separate process, they, generally, can't block UI just by doing a blocking call
.